src/languages/reasonml.js

```diff
@@ -6,301 +6,63 @@ Author: Gidi Meir Morris <oss@gidi.io>
+        + 'exception external for fun function functor if in include inherit '
+        + 'initializer land lazy let lor lsl lsr lxor mod module mutable new '
+        + 'nonrec object of open or pri pub rec sig struct switch then to try '
+        + 'type val virtual when while with',
+      built_in:
+        'array bool bytes char exn|5 float int int32 int64 list lazy_t|5 nativeint|5 ref string unit ',
+      literal:
+        'true false'
+    },
     illegal: '(:-|:=|\\$\\{|\\+=)',
     contains: [
-      hljs.COMMENT('/\\*', '\\*/', { illegal: '^(#,\\/\\/)' }),
-      {
-        className: 'character',
-        begin: '\'(\\\\[^\']+|[^\'])\'',
-        illegal: '\\n',
-        relevance: 0
-      },
-      hljs.QUOTE_STRING_MODE,
       {
         className: 'literal',
-        begin: '\\(\\)',
+        begin: '\\[(\\|\\|)?\\]|\\(\\)',
         relevance: 0
       },
-      {
-        className: 'literal',
-        begin: '\\[\\|',
-        end: '\\|\\]',
-        relevance: 0,
-        contains: LIST_CONTENTS_MODES
+      hljs.C_LINE_COMMENT_MODE,
+      hljs.COMMENT('/\\*', '\\*/', { illegal: '^(#,\\/\\/)' }),
+      { /* type variable */
+        className: 'symbol',
+        begin: '\'[A-Za-z_](?!\')[\\w\']*'
+        /* the grammar is ambiguous on how 'a'b should be interpreted but not the compiler */
       },
-      {
-        className: 'literal',
-        begin: '\\[',
-        end: '\\]',
-        relevance: 0,
-        contains: LIST_CONTENTS_MODES
+      { /* polymorphic variant */
+        className: 'type',
+        begin: '`[A-Z][\\w\']*'
       },
-      CONSTRUCTOR_MODE,
-      {
-        className: 'operator',
-        begin: RE_OPERATOR_SPACED,
-        illegal: '-->',
+      { /* module or constructor */
+        className: 'type',
+        begin: '\\b[A-Z][\\w\']*',
         relevance: 0
       },
-      NUMBER_MODE,
-      hljs.C_LINE_COMMENT_MODE,
-      PATTERN_MATCH_BLOCK_MODE,
-      FUNCTION_BLOCK_MODE,
+      { /* don't color identifiers, but safely catch all identifiers with ' */
+        begin: '[a-z_]\\w*\'[\\w\']*',
+        relevance: 0
+      },
+      hljs.inherit(hljs.APOS_STRING_MODE, {
+        className: 'string',
+        relevance: 0
+      }),
+      hljs.inherit(hljs.QUOTE_STRING_MODE, { illegal: null }),
       {
-        className: 'module-def',
-        begin: "\\bmodule\\s+" + RE_IDENT + "\\s+" + RE_MODULE_IDENT + "\\s+=\\s+\\{",
-        end: /\}/,
-        returnBegin: true,
-        keywords: KEYWORDS,
-        relevance: 0,
-        contains: [
-          {
-            className: 'module',
-            relevance: 0,
-            begin: RE_MODULE_IDENT
-          },
-          {
-            begin: /\{/,
-            end: /\}/,
-            relevance: 0,
-            skip: true
-          }
-        ].concat(MODULE_ACCESS_CONTENTS)
+        className: 'number',
+        begin:
+          '\\b(0[xX][a-fA-F0-9_]+[Lln]?|'
```

Comment:
Lets use the regex tools (`hljs.regex`) here... either `regex.either` or this could also be done with `variants` (see docs).

Reply:
https://github.com/highlightjs/highlight.js/pull/3795/commits/355e5003525e5ef24799a6167372606e29c1c5f5